stdlib/public/core/FloatingPointTypes.swift.gyb

```diff
@@ -1599,17 +1608,26 @@ extension ${Self} {
   /// can't be represented exactly, the result is `nil`.
   ///
   /// - Parameter value: The integer to represent as a floating-point value.
-%   if srcBits < SignificandBitCount:
+%   if srcBits <= dstBits:
   @available(*, message: "Converting ${That} to ${Self} will always succeed.")
 %   end
   @inlinable // FIXME(sil-serialize-all)
   @inline(__always)
   public init?(exactly value: ${That}) {
     _value = Builtin.${sign}itofp_${ThatBuiltinName}_FPIEEE${bits}(value._value)
 
-%   if srcBits >= SignificandBitCount:
-    guard let roundTrip = ${That}(exactly: self),
-      roundTrip == value else {
+%   if srcBits > dstBits:
+    // The magnitude of the integer must be representable by the significand.
+    // Omit trailing zeros that can be encoded by the exponent.
+    let absv = value.magnitude
+    let mask: ${That}.Magnitude = ~((1 << (${SignificandBitCount} + 1)) - 1)
+    let tz = absv.trailingZeroBitCount
+    if _slowPath((absv >> tz) & mask != 0) {
+      return nil
+    }
+    // Check that the required exponent is in range.
+    let maxexp = (1 << (${ExponentBitCount} - 1)) - 1
```

Comment:
It'd be nice if we could tighten the GYB condition not to assume that the maximum exponent is representable as an integer. If this overflows, we'll erroneously return `nil` when in fact the maximum exponent is more than enough to accommodate the integer.

Reply:
The code here only covers Float and Double currently however (the LLVM intrinsic used for the conversion also relies on this), so we should only have 8 or 11 bits of exponent. If it's alright with you I'll just add a check to make sure we won't overflow.